File: sdk/go/internal/crypto/address.go
package crypto

import (
	"crypto/sha256"
	"fmt"

	"golang.org/x/crypto/ripemd160"
)

const (
	// Bech32PrefixAccAddr is the Bech32 prefix for account addresses.
	Bech32PrefixAccAddr = "seocheon1"
)

// AddressFromPubKey derives a Cosmos bech32 address from a compressed public key.
func AddressFromPubKey(pubKey []byte) (string, error) {
	if len(pubKey) != 33 {
		return "", fmt.Errorf("public key must be 33 bytes (compressed), got %d", len(pubKey))
	}

	// SHA256 hash
	sha := sha256.Sum256(pubKey)

	// RIPEMD160 hash
	rip := ripemd160.New()
	rip.Write(sha[:])
	addrBytes := rip.Sum(nil) // 20 bytes

	// Bech32 encode with "seocheon" HRP (produces "seocheon1..." address)
	bech32Addr, err := bech32Encode("seocheon", addrBytes)
	if err != nil {
		return "", fmt.Errorf("bech32 encoding: %w", err)
	}

	return bech32Addr, nil
}

// bech32Encode encodes data to bech32 format.
// This is a minimal implementation for Cosmos address encoding.
func bech32Encode(hrp string, data []byte) (string, error) {
	// Convert 8-bit data to 5-bit groups
	converted, err := convertBits(data, 8, 5, true)
	if err != nil {
		return "", fmt.Errorf("converting bits: %w", err)
	}

	// Calculate checksum
	values := append(converted, 0, 0, 0, 0, 0, 0)
	polymod := bech32Polymod(expandHRP(hrp, values)) ^ 1
	for i := 0; i < 6; i++ {
		values[len(converted)+i] = byte((polymod >> uint(5*(5-i))) & 31)
	}

	// Encode to string
	charset := "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
	result := hrp + "1"
	for _, v := range values {
		result += string(charset[v])
	}

	return result, nil
}

// convertBits converts a byte slice from one bit-group size to another.
func convertBits(data []byte, fromBits, toBits uint, pad bool) ([]byte, error) {
	acc := uint32(0)
	bits := uint(0)
	var result []byte
	maxV := uint32((1 << toBits) - 1)

	for _, b := range data {
		acc = (acc << fromBits) | uint32(b)
		bits += fromBits
		for bits >= toBits {
			bits -= toBits
			result = append(result, byte((acc>>bits)&maxV))
		}
	}

	if pad {
		if bits > 0 {
			result = append(result, byte((acc<<(toBits-bits))&maxV))
		}
	} else if bits >= fromBits {
		return nil, fmt.Errorf("illegal zero padding")
	} else if ((acc << (toBits - bits)) & maxV) != 0 {
		return nil, fmt.Errorf("non-zero padding")
	}

	return result, nil
}

// expandHRP expands the human-readable part for checksum computation.
func expandHRP(hrp string, values []byte) []byte {
	result := make([]byte, 0, len(hrp)*2+1+len(values))
	for _, c := range hrp {
		result = append(result, byte(c>>5))
	}
	result = append(result, 0)
	for _, c := range hrp {
		result = append(result, byte(c&31))
	}
	result = append(result, values...)
	return result
}

// bech32Polymod computes the bech32 checksum.
func bech32Polymod(values []byte) uint32 {
	gen := [5]uint32{0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3}
	chk := uint32(1)
	for _, v := range values {
		top := chk >> 25
		chk = (chk&0x1ffffff)<<5 ^ uint32(v)
		for i := 0; i < 5; i++ {
			if (top>>uint(i))&1 == 1 {
				chk ^= gen[i]
			}
		}
	}
	return chk
}

File: sdk/csharp/src/Internal/Crypto/Address.cs
using System.Security.Cryptography;
using Seocheon.Sdk.Constants;
using Seocheon.Sdk.Errors;

namespace Seocheon.Sdk.Internal.Crypto;

/// <summary>
/// Cosmos address derivation: pubkey → SHA256 → RIPEMD160 → bech32("seocheon").
/// </summary>
public static class Address
{
    /// <summary>
    /// Derives a bech32 address from a compressed public key (33 bytes).
    /// </summary>
    public static string FromPubKey(byte[] pubKey)
    {
        if (pubKey.Length != 33)
            throw SdkErrors.InvalidAddress("Public key must be 33 bytes (compressed)");

        // SHA256
        var sha256Hash = SHA256.HashData(pubKey);

        // RIPEMD160
        var ripemd = new Org.BouncyCastle.Crypto.Digests.RipeMD160Digest();
        ripemd.BlockUpdate(sha256Hash, 0, sha256Hash.Length);
        var ripemdHash = new byte[20];
        ripemd.DoFinal(ripemdHash, 0);

        // Bech32 encode
        return Bech32Encode(ChainConstants.AddressPrefix, ripemdHash);
    }

    /// <summary>
    /// Validates a bech32 address with the seocheon prefix.
    /// </summary>
    public static bool Validate(string address)
    {
        if (string.IsNullOrWhiteSpace(address))
            return false;

        try
        {
            var (hrp, data) = Bech32Decode(address);
            return hrp == ChainConstants.AddressPrefix && data.Length == 20;
        }
        catch
        {
            return false;
        }
    }

    // === Bech32 Implementation ===

    private const string Charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";

    private static readonly int[] Generator = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];

    private static int Polymod(int[] values)
    {
        var chk = 1;
        foreach (var v in values)
        {
            var b = chk >> 25;
            chk = ((chk & 0x1ffffff) << 5) ^ v;
            for (var i = 0; i < 5; i++)
            {
                if (((b >> i) & 1) != 0)
                    chk ^= Generator[i];
            }
        }
        return chk;
    }

    private static int[] HrpExpand(string hrp)
    {
        var result = new int[hrp.Length * 2 + 1];
        for (var i = 0; i < hrp.Length; i++)
        {
            result[i] = hrp[i] >> 5;
            result[i + hrp.Length + 1] = hrp[i] & 31;
        }
        result[hrp.Length] = 0;
        return result;
    }

    private static bool VerifyChecksum(string hrp, int[] data)
    {
        var values = HrpExpand(hrp).Concat(data).ToArray();
        return Polymod(values) == 1;
    }

    private static int[] CreateChecksum(string hrp, int[] data)
    {
        var values = HrpExpand(hrp).Concat(data).Concat(new int[6]).ToArray();
        var polymod = Polymod(values) ^ 1;
        var checksum = new int[6];
        for (var i = 0; i < 6; i++)
            checksum[i] = (polymod >> (5 * (5 - i))) & 31;
        return checksum;
    }

    private static byte[] ConvertBits(byte[] data, int fromBits, int toBits, bool pad)
    {
        var acc = 0;
        var bits = 0;
        var result = new List<byte>();
        var maxv = (1 << toBits) - 1;

        foreach (var value in data)
        {
            acc = (acc << fromBits) | value;
            bits += fromBits;
            while (bits >= toBits)
            {
                bits -= toBits;
                result.Add((byte)((acc >> bits) & maxv));
            }
        }

        if (pad)
        {
            if (bits > 0)
                result.Add((byte)((acc << (toBits - bits)) & maxv));
        }
        else if (bits >= fromBits || ((acc << (toBits - bits)) & maxv) != 0)
        {
            throw new FormatException("Invalid padding in bech32 conversion");
        }

        return result.ToArray();
    }

    internal static string Bech32Encode(string hrp, byte[] data)
    {
        var converted = ConvertBits(data, 8, 5, true);
        var intData = converted.Select(b => (int)b).ToArray();
        var checksum = CreateChecksum(hrp, intData);
        var combined = intData.Concat(checksum).ToArray();

        var result = hrp + "1";
        foreach (var v in combined)
            result += Charset[v];

        return result;
    }

    internal static (string hrp, byte[] data) Bech32Decode(string bech)
    {
        bech = bech.ToLowerInvariant();
        var pos = bech.LastIndexOf('1');
        if (pos < 1 || pos + 7 > bech.Length)
            throw new FormatException("Invalid bech32 string");

        var hrp = bech[..pos];
        var dataChars = bech[(pos + 1)..];
        var data = new int[dataChars.Length];

        for (var i = 0; i < dataChars.Length; i++)
        {
            var idx = Charset.IndexOf(dataChars[i]);
            if (idx < 0)
                throw new FormatException($"Invalid bech32 character: {dataChars[i]}");
            data[i] = idx;
        }

        if (!VerifyChecksum(hrp, data))
            throw new FormatException("Invalid bech32 checksum");

        // Remove checksum (last 6 values)
        var payload = data[..^6].Select(v => (byte)v).ToArray();
        var decoded = ConvertBits(payload, 5, 8, false);

        return (hrp, decoded);
    }
}
